File: node-backend/utils/geminiAI.js
/**
 * geminiAI.js — Production AI engine using Google Gemini
 *
 * Model: gemini-2.0-flash (fast, smart, free tier: 1,500 req/day)
 * Cascade: Gemini → Groq → Pollinations → Local fallback
 *
 * Key format: AQ.Ab8... (new Google format, works the same as AIzaSy...)
 */

import { callGroq } from './groqAI.js';

const GEMINI_MODEL = 'gemini-2.0-flash';
const GEMINI_BASE  = 'https://generativelanguage.googleapis.com/v1beta/models';

function getKey() {
    const k = process.env.VITE_GEMINI_API_KEY;
    return (k && k !== 'your_new_gemini_api_key_here' && k.trim().length > 10) ? k.trim() : null;
}

// ── Core Gemini call ──────────────────────────────────────────────────────────
async function callGemini(systemPrompt, userMessage, jsonMode = false) {
    const key = getKey();
    if (!key) { console.log('[Gemini] No key configured'); return null; }

    try {
        const url = `${GEMINI_BASE}/${GEMINI_MODEL}:generateContent?key=${key}`;
        const body = {
            systemInstruction: { parts: [{ text: systemPrompt }] },
            contents: [{ role: 'user', parts: [{ text: userMessage }] }],
            generationConfig: {
                temperature: jsonMode ? 0.2 : 0.7,
                maxOutputTokens: jsonMode ? 8192 : 4096,
                ...(jsonMode && { responseMimeType: 'application/json' }),
            },
        };

        const res = await fetch(url, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(body),
            signal: AbortSignal.timeout(10000),
        });

        if (res.ok) {
            const data = await res.json();
            const text = data?.candidates?.[0]?.content?.parts?.[0]?.text?.trim();
            if (text && text.length > 0) {
                console.log(`[Gemini] ✓ served (${text.length} chars)`);
                return text;
            }
        } else {
            const err = await res.json().catch(() => ({}));
            const msg = err?.error?.message?.slice(0, 100) || res.status;
            console.warn(`[Gemini] HTTP ${res.status}: ${msg}`);
            // Quota/auth errors won't recover on retry — skip slow fallbacks in cascade
            if (res.status === 429 || res.status === 403 || res.status === 401) {
                return '__GEMINI_SKIP__';
            }
        }
    } catch (err) {
        console.warn('[Gemini] request failed:', err.message?.slice(0, 80));
    }
    return null;
}

// ── Pollinations fallback ─────────────────────────────────────────────────────
async function callPollinations(systemPrompt, userMessage) {
    // Try multiple Pollinations endpoints for reliability
    const endpoints = [
        {
            url: 'https://text.pollinations.ai/',
            body: {
                messages: [
                    { role: 'system', content: systemPrompt },
                    { role: 'user',   content: userMessage  },
                ],
                model: 'openai',
                seed: 42,
            },
        },
        {
            url: 'https://text.pollinations.ai/',
            body: {
                messages: [
                    { role: 'system', content: systemPrompt },
                    { role: 'user',   content: userMessage  },
                ],
                model: 'mistral',
            },
        },
    ];

    for (const ep of endpoints) {
        try {
            const res = await fetch(ep.url, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json', 'User-Agent': 'Mozilla/5.0' },
                body: JSON.stringify(ep.body),
                signal: AbortSignal.timeout(8000),
            });
            if (res.ok) {
                const text = (await res.text()).trim();
                if (text.length > 10) {
                    console.log(`[Pollinations] ✓ fallback served (${text.length} chars)`);
                    return text;
                }
            }
        } catch (err) {
            console.warn('[Pollinations] endpoint failed:', err.message?.slice(0, 60));
        }
    }
    return null;
}

// ── Public API ────────────────────────────────────────────────────────────────

/** Chat: Gemini → Groq → Pollinations → null (caller handles local fallback) */
export async function aiChat(systemPrompt, userMessage) {
    const gemini = await callGemini(systemPrompt, userMessage, false);
    if (gemini && gemini !== '__GEMINI_SKIP__') return gemini;

    const groq = await callGroq(systemPrompt, userMessage, false);
    if (groq) return groq;

    return await callPollinations(systemPrompt, userMessage) || null;
}

/**
 * JSON generation for structured content (PDF, PPT, avatar prompt).
 * Returns parsed JS object or null.
 */
export async function aiJsonGenerate(systemPrompt, userPrompt) {
    // 1. Try Gemini with JSON mode
    let raw = await callGemini(systemPrompt, userPrompt, true);
    if (raw) {
        try { return JSON.parse(raw); } catch (_) { /* fall through */ }
    }

    // 2. Try Pollinations (plain text, parse manually)
    raw = await callPollinations(systemPrompt, userPrompt);
    if (raw) {
        let cleaned = raw.trim().replace(/^```[a-z]*\n?/, '').replace(/```$/, '').trim();
        try { return JSON.parse(cleaned); } catch (_) { /* fall through */ }
    }

    return null; // caller uses local template
}

export const isGeminiConfigured = () => !!getKey();

File: backend/gemini_ai.py
"""
gemini_ai.py — Production AI engine using Google Gemini 2.0 Flash
Cascade: Gemini → Pollinations (free, no key) → None

Free tier: 1,500 requests/day, no credit card
Model: gemini-2.0-flash (fast + smart)
"""

import os
import json
import re
import requests

GEMINI_MODEL = "gemini-2.0-flash"
GEMINI_BASE  = "https://generativelanguage.googleapis.com/v1beta/models"
POLLINATIONS_URL = "https://text.pollinations.ai/"


def _get_key() -> str | None:
    k = os.getenv("VITE_GEMINI_API_KEY", "").strip()
    return k if k and k != "your_new_gemini_api_key_here" and len(k) > 10 else None


def _pollinations_chat(system_prompt: str, user_message: str, timeout: int = 20) -> str | None:
    """Free Pollinations fallback — no API key needed."""
    try:
        resp = requests.post(
            POLLINATIONS_URL,
            headers={"Content-Type": "application/json", "User-Agent": "Mozilla/5.0"},
            json={
                "messages": [
                    {"role": "system", "content": system_prompt},
                    {"role": "user",   "content": user_message},
                ],
            },
            timeout=timeout,
        )
        if resp.status_code == 200:
            text = resp.text.strip()
            if text:
                print(f"[Pollinations] ✓ fallback served ({len(text)} chars)")
                return text
    except Exception as e:
        print(f"[Pollinations] error: {e}")
    return None


def gemini_chat(system_prompt: str, user_message: str, timeout: int = 25) -> str | None:
    """
    Send a chat message to Gemini → Pollinations fallback.
    Returns text or None on complete failure.
    """
    key = _get_key()
    if key:
        url = f"{GEMINI_BASE}/{GEMINI_MODEL}:generateContent?key={key}"
        body = {
            "systemInstruction": {"parts": [{"text": system_prompt}]},
            "contents": [{"role": "user", "parts": [{"text": user_message}]}],
            "generationConfig": {
                "temperature": 0.7,
                "maxOutputTokens": 4096,
            },
        }
        try:
            resp = requests.post(url, json=body, timeout=timeout)
            if resp.status_code == 200:
                text = resp.json()["candidates"][0]["content"]["parts"][0]["text"].strip()
                if text:
                    print(f"[Gemini] ✓ chat served ({len(text)} chars)")
                    return text
            else:
                err = resp.json().get("error", {}).get("message", "unknown")
                print(f"[Gemini] HTTP {resp.status_code}: {err[:100]}")
        except Exception as e:
            print(f"[Gemini] chat error: {e}")

    # Pollinations fallback
    return _pollinations_chat(system_prompt, user_message)


def gemini_json(system_prompt: str, user_prompt: str, timeout: int = 30) -> dict | None:
    """
    Call Gemini expecting a JSON response.
    Returns parsed dict or None on failure.
    """
    key = _get_key()
    if key:
        url = f"{GEMINI_BASE}/{GEMINI_MODEL}:generateContent?key={key}"
        body = {
            "systemInstruction": {"parts": [{"text": system_prompt}]},
            "contents": [{"role": "user", "parts": [{"text": user_prompt}]}],
            "generationConfig": {
                "temperature": 0.2,
                "maxOutputTokens": 8192,
                "responseMimeType": "application/json",
            },
        }
        try:
            resp = requests.post(url, json=body, timeout=timeout)
            if resp.status_code == 200:
                raw = resp.json()["candidates"][0]["content"]["parts"][0]["text"].strip()
                if raw.startswith("```"):
                    raw = re.sub(r"^```[a-z]*\n?", "", raw).rstrip("`").strip()
                result = json.loads(raw)
                print(f"[Gemini] ✓ JSON served")
                return result
            else:
                err = resp.json().get("error", {}).get("message", "unknown")
                print(f"[Gemini] JSON HTTP {resp.status_code}: {err[:100]}")
        except Exception as e:
            print(f"[Gemini] JSON error: {e}")

    # Try Pollinations for JSON too
    try:
        raw = _pollinations_chat(system_prompt + "\n\nIMPORTANT: respond with valid JSON only, no markdown.", user_prompt)
        if raw:
            cleaned = re.sub(r"^```[a-z]*\n?", "", raw.strip()).rstrip("`").strip()
            return json.loads(cleaned)
    except Exception:
        pass

    return None


def is_gemini_configured() -> bool:
    return _get_key() is not None
